Deprecate normals and fitting

briandleahy opened the pull request:
This small PR deprecates normals and fitting, as part of #394 . 

I will split the deprecations to nmpfit out in a different PR, so we can keep things modular. This PR is modular with respect to nmpfit; as far as I can tell it will have no effects on the nmpfit machinery.

pep8speaks commented:
Hello @briandleahy! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`holopy/core/errors.py`](https://github.com/manoharan-lab/holopy/blob/105c93c0421b2f7ca69be10838335f7af530a2e7/holopy/core/errors.py):

> [Line 66:5](https://github.com/manoharan-lab/holopy/blob/105c93c0421b2f7ca69be10838335f7af530a2e7/holopy/core/errors.py#L66): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

* In the file [`holopy/core/io/io.py`](https://github.com/manoharan-lab/holopy/blob/105c93c0421b2f7ca69be10838335f7af530a2e7/holopy/core/io/io.py):

> [Line 524:80](https://github.com/manoharan-lab/holopy/blob/105c93c0421b2f7ca69be10838335f7af530a2e7/holopy/core/io/io.py#L524): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)

barkls commented on holopy/inference/result.py:
```diff
@@ -187,13 +188,13 @@ def _save(self, filename, **kwargs):
         # this method is published in the HoloPy paper
-        from holopy.fitting import fit_warning
-        fit_warning('FitResult.hologram', 'SamplingResult.best_fit()')
+        raise_fitting_api_error(
```
I feel like any previously published code should be held onto for as long as possible in terms of either making it work, or providing an informative message (like you've done here), so I think it's good to keep this around for now.

barkls reviewed (commented):
Thanks for this!

I think we should do a (major) 3.5 when all of our changes are ready and remove as much deprecated content as possible. It's not clear when the next development cycle for HoloPy will be or who would work on it, so my opinion is that we should try to make 3.5 as much of a stable, finished product as possible. This means cutting out as much as possible that is old, even if it hasn't been deprecated very long. Let me know what your thoughts are!

You should update the releasenotes regarding these changes. If you follow my recommendation and break backwards compatibility of saving files, it might also be worth putting a disclaimer somewhere prominent that the 3.4 release can convert old files to the new format.

briandleahy commented on holopy/inference/model.py:
```diff
@@ -46,10 +47,11 @@ def _purge_ties(node):
             tie_msg = ('. Ignoring previously defined ties. '
                         'Use Model.add_tie() to reassign them')
-            fit_warning('newly saved model', 'the old one' + tie_msg)
+            raise_fitting_api_error(
```
deleted

briandleahy commented on holopy/core/errors.py:
```diff
@@ -50,10 +47,21 @@ class DependencyMissing(Exception):
+    pass
+
+
+def raise_fitting_api_error(correct_obj, obselete_obj):
```
I've kept it for the 1 bit of code in the holopy paper that we are deprecating (result.best_fit).

briandleahy commented:
I went on a spree and deleted basically everything related to the inference that could be deprecated (with the exception of nmpfit, which I will do later). I left in the `result.best_fit` method in with the same descriptive error message, since it is published in the holopy paper. I've also updated the release notes. 

Let me know your thoughts.

barkls commented on holopy/inference/model.py:
```diff
@@ -535,20 +510,6 @@ def _lnlike(self, pars, data):
             0.5 * (self._residuals(pars, data, noise_sd)**2).sum())
         return log_likelihood
 
-    def fit(self, data, strategy=None):
```
I think you could have `model.fit()` and `model.sample()` methods `raise_fitting_api_error`. My rationale is that they don't really require any other adjustments and it's easy for someone to delete the few lines of code later so the cost of keeping them is minimal and the benefit is better clarity for anybody trying to use them but it's up to you.
